HIP/hip-729.md

```diff
@@ -0,0 +1,129 @@
+
+## Rationale
+
+As described in [HIP-482](https://hips.hedera.com/hip/hip-482), the `JSON-RPC Relay` works with `EthereumTransactions` that it forwards to the consensus nodes. Apart from the `EthereumTransaction` type, the consensus nodes can also handle other types of transactions (e.g. `ContractCall` or `ContractCreate`) that do not come from the `JSON-RPC Relay` but can be used by developers by other means.
+In general, when we talk about account nonces we will need to consider all of those transaction types for both types of accounts (i.e. EOAs and contracts):
+
+- Contract nonce updates for both cases of `EthereumTransactions` and `non-EthereumTransactions` (i.e. `ContractCall` or `ContractCreate`) will need to be tracked and externalized to Mirror Node
+- Tracking nonces for EOAs for `EthereumTransactions` is not changed (currently we don't have a specific protobuf to externalize the nonce but the nonce value is passed as part of the `EthereumTransaction` body, so Mirror Node can read the value from there)
+- Tracking nonces for EOAs for `non-EthereumTransactions` is not supported because there is no use case for it in Hedera
+
+## User stories
+
+1. As a user, when I make a request to the Mirror Node `api/v1/contracts/{contractId}` I observe a valid non-zero contract nonce value.
+2. As a user, when I make a request to `eth_getTransactionCount` endpoint I want to be able to get correct nonce values for contracts.
+  
+## Specification
+
+No changes to Services state will be needed because currently the nonces are currently being computed correctly there. We need to externalize these changes to Mirror Node so that it can keep a correct state representation for contract accounts.
+
+### HAPI Changes
+
+Define new message type:
+
+```protobuf
+message ContractNonceInfo {
+
+    /**
+     * Id of the contract
+     */
+    ContractID contract_id = 1;
+
+    /**
+     * The current value of the contract account's nonce property
+     */
+    int64 nonce = 2;
+}
+```
+
+Update `ContractFunctionResult`:
+
+```protobuf
+message ContractFunctionResult {
+    ...
+
+    /**
+     * A list of updated contract account nonces containing the new nonce value for each contract account
+     */
+    repeated ContractNonceInfo contract_nonces = 14;
+}
+```
+
+### Mirror Node Changes
+
+Mirror Node should start using the new `contract_nonces` field and update the state of the contract accounts with it.  Additionally, the `/api/v1/contracts` and `api/v1/contract/{contractIdOrAddress}` endpoints responses should be updated to return the contract `nonce` value:
+```json
+{
+  ...,
+  "nonce": 1,
+  ...
+}
+```
+
```

Comment:
Please incorporate @SimiHunjan 's suggestion to add a section for `SDK` and `Mirror Node Explorer` changes